src/rational/mat_q/cholesky_decomp.rs

```diff
@@ -0,0 +1,151 @@
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module implements the Cholesky Decomposition for Hermitian positive-definite
+//! matrices.
+
+use super::MatQ;
+use crate::{
+    rational::Q,
+    traits::{Concatenate, GetEntry, GetNumColumns},
+};
+
+impl MatQ {
+    /// This function performs the Cholesky decomposition and produces a matrix `L`
+    /// such that `self = L * L^T`.
+    /// This function relies on the precision of `Q::sqrt` and will not provide
+    /// perfect results in all cases.
+    /// Furthermore, the Cholesky decomposition requires a Hermitian positive-definite
+    /// matrix.
+    /// This function does not check whether such a matrix is provided and unintentional
+    /// behaviors or outputs may occur if the provided matrix is not positive-definite.
+    ///
+    /// Returns the Cholesky decomposition of a Hermitian positive-definite matrix.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    /// use std::str::FromStr;
+    ///
+    /// let matrix = MatQ::from_str("[[4, 12, -16],[12,37,-43],[-16,-43,98]]").unwrap();
+    ///
+    /// let l = matrix.cholesky_decomposition();
+    /// assert_eq!(matrix, &l * l.transpose());
+    /// ```
+    ///
+    /// # Panics ...
+    /// - if `self` is not a symmetric matrix.
+    ///
+    /// # Note
+    /// - the function may behave unintended if the provided matrix is not a
+    ///     Hermitian positive-definite matrix.
+    pub fn cholesky_decomposition(&self) -> MatQ {
+        assert!(self.is_symmetric(), "The provided matrix is not symmetric.");
+        // TODO: replace manual implementation with faster implementation from
+        // FLINT directly, once that is accessible through the FFI
+        let n = self.get_num_columns();
+
+        let mut a = self.clone();
+        let mut l = MatQ::identity(n, n);
+
+        for i in 0..n {
+            // get first entry and ith column
+            let a_ii = a.get_entry(0, 0).unwrap();
+            assert!(a_ii > Q::ZERO, "The matrix is not positive-definite.");
+            let column_a_i = match i {
+                0 => a.get_column(0).unwrap(),
+                _ => MatQ::new(i, 1)
+                    .get_column(0)
+                    .unwrap()
+                    .concat_vertical(&a.get_column(0).unwrap())
+                    .unwrap(),
+            } * (1 / (a_ii.sqrt()));
+
+            // produce L matrix recursively
+            let mut l_i = MatQ::identity(n, n);
+            l_i.set_column(i, &column_a_i, 0).unwrap();
+            l = l * l_i;
+
+            // update matrix A recursively
+            if i < n - 1 {
+                let b = a.get_submatrix(1, n - i - 1, 1, n - i - 1).unwrap();
+                let b_minus = (1 / a_ii)
+                    * a.get_submatrix(1, n - i - 1, 0, 0).unwrap()
+                    * a.get_submatrix(0, 0, 1, n - i - 1).unwrap();
```

Comment:
Matrix is symmetric, is transpose faster than get_submatrix?

Reply:
if one of them is faster, then it should be get_submatrix - I think.

Both utilize the underlying Flint-functions: transpose and window respectively.

transpose iterates through all entries and sets the entries respectively - therefore we have the additional iteration.
the window first gets pointers for each entry and then simply performs a deepclone of all entries in the range.

In the end -> both should be basically identical